Migrate `pdb-buildinfo-cl-cmd` and `pgo-indirect-call-promotion` `run-make` tests to rmake

Oneirical opened the pull request:
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

Please try:

try-job: x86_64-msvc
try-job: x86_64-mingw
try-job: i686-msvc
try-job: i686-mingw
try-job: x86_64-gnu-llvm-17
try-job: aarch64-apple

Oneirical committed:
rewrite pdb-buildinfo-cl-cmd to rmake

Oneirical committed:
rewrite pgo-indirect-call-promotion to rmake

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit c8669591fcdd8abb9493c28e1cb7d1a6a92ecc85 with merge 0c4a7a49b8a03e439c3d4665c76138165c7defaa...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c8669591fcdd8abb9493c28e1cb7d1a6a92ecc85","merge_sha":"0c4a7a49b8a03e439c3d4665c76138165c7defaa"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10161563342/job/28100314843) [(plain)](https://github.com/rust-lang-ci/rust/commit/0c4a7a49b8a03e439c3d4665c76138165c7defaa/checks/28100314843/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  99% (48376/48864)
Updating files:  99% (48480/48864)
Updating files: 100% (48864/48864)
Updating files: 100% (48864/48864), done.
branch 'try' set up to track 'origin/try'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
'0c4a7a49b8a03e439c3d4665c76138165c7defaa'
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
---- [run-make] tests\run-make\pdb-buildinfo-cl-cmd stdout ----

error: rmake recipe failed to complete
status: exit code: 101
command: "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-make\\pdb-buildinfo-cl-cmd\\rmake.exe"
--- stderr -------------------------------
thread 'main' panicked at C:\a\rust\rust\tests\run-make\pdb-buildinfo-cl-cmd\rmake.rs:20:21:
thread 'main' panicked at C:\a\rust\rust\tests\run-make\pdb-buildinfo-cl-cmd\rmake.rs:20:21:
the file in path "my_crate_name.pdb" could not be read into a String: Error { kind: InvalidData, message: "stream did not contain valid UTF-8" }
------------------------------------------




failures:
    [run-make] tests\run-make\pdb-buildinfo-cl-cmd

test result: FAILED. 261 passed; 1 failed; 90 ignored; 0 measured; 0 filtered out; finished in 83.06s

Some tests failed in compiletest suite=run-make mode=run-make host=x86_64-pc-windows-msvc target=x86_64-pc-windows-msvc
Build completed unsuccessfully in 1:29:35
make: *** [Makefile:102: ci-msvc-ps1] Error 1
  network time: Tue, 30 Jul 2024 13:14:40 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

Oneirical commented:
Invalid UTF-8 error. Maybe this will fare better, but I have doubts.

`assert_contains(String::from_utf8_lossy(&rfs::read("my_crate_name.pdb")), env_var("RUSTC"));`

@rustbot ready

Please run try again, if you think this stands a shot.

jieyouxu commented:
> Invalid UTF-8 error. Maybe this will fare better, but I have doubts.
> 
> `assert_contains(String::from_utf8_lossy(&rfs::read("my_crate_name.pdb")), env_var("RUSTC"));`

Shouldn't we use something like `assert_invalid_utf8_contains`, because AFAIK pdb debug info files don't need to be UTF-8 (and it wouldn't make sense if it did).

ChrisDenton commented:
Yeah. PDB is a binary format so it's highly unlikely to be valid UTF-8 and doing a lossy conversion seems like a lot of unnecessary work. Basically this should read the file as bytes (`fs::read`) then search within it using `env_var("RUSTC").as_bytes()` or some helper that does that.

jieyouxu commented:
Yeah we could use something like [`bstr::ByteSlice::find_iter`](https://docs.rs/bstr/latest/bstr/trait.ByteSlice.html#method.find_iter).

@rustbot author

bors commented:
:umbrella: The latest upstream changes (presumably #128075) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Oneirical commented:
Okay, I gave the `bstr` implementation a try, but I am of little faith - I tried to run this locally without the `.pdb` extension on a Linux system and failed at the RUSTC assertion. It's possible that the `pdb` file on MSVC is different and that this is the whole point of the test, though.

@rustbot review

jieyouxu commented on tests/run-make/pdb-buildinfo-cl-cmd/rmake.rs:
```diff
@@ -0,0 +1,31 @@
+        .metadata("dc9ef878b0a48666")
+        .run();
+    use bstr::ByteSlice;
+    assert!(&rfs::read("my_crate_name.pdb").find(env_var("RUSTC").as_bytes()).is_some());
+    assert!(&rfs::read("my_crate_name.pdb").find(br#""main.rs""#).is_some());
+    assert!(&rfs::read("my_crate_name.pdb").find(br#""-g""#).is_some());
+    assert!(&rfs::read("my_crate_name.pdb").find(br#""--crate-name""#).is_some());
+    assert!(&rfs::read("my_crate_name.pdb").find(br#""my_crate_name""#).is_some());
+    assert!(&rfs::read("my_crate_name.pdb").find(br#""--crate-type""#).is_some());
+    assert!(&rfs::read("my_crate_name.pdb").find(br#""bin""#).is_some());
+    assert!(&rfs::read("my_crate_name.pdb").find(br#""-C""#).is_some());
+    assert!(&rfs::read("my_crate_name.pdb").find(br#""metadata=dc9ef878b0a48666""#).is_some());
+    assert!(&rfs::read("my_crate_name.pdb").find(br#""--out-dir""#).is_some());
```
Suggestion: a helper like `assert_pdb_contains`, probably just in this test file.

jieyouxu commented:
> Okay, I gave the `bstr` implementation a try, but I am of little faith - I tried to run this locally without the `.pdb` extension on a Linux system and failed at the RUSTC assertion. It's possible that the `pdb` file on MSVC is different and that this is the whole point of the test, though.

Isn't pdb files a microsoft concept?

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit 76c5149c757e6493f9e5a38d83ab1b6ecefe46d1 with merge 142ab629b7073e6af2e2f84bf0ad87604ff659fa...
<!-- homu: {"type":"TryBuildStarted","head_sha":"76c5149c757e6493f9e5a38d83ab1b6ecefe46d1","merge_sha":"142ab629b7073e6af2e2f84bf0ad87604ff659fa"} -->

rust-log-analyzer commented:
The job **`x86_64-msvc`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/10224700120/job/28292784412) [(plain)](https://github.com/rust-lang-ci/rust/commit/142ab629b7073e6af2e2f84bf0ad87604ff659fa/checks/28292784412/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Updating files:  99% (48422/48911)
Updating files:  99% (48854/48911)
Updating files: 100% (48911/48911)
Updating files: 100% (48911/48911), done.
branch 'try' set up to track 'origin/try'.
Switched to a new branch 'try'
[command]"C:\Program Files\Git\bin\git.exe" log -1 --format='%H'
'142ab629b7073e6af2e2f84bf0ad87604ff659fa'
##[group]Run src/ci/scripts/setup-environment.sh
src/ci/scripts/setup-environment.sh
---
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
file:.git/config submodule.library/stdarch.url=https://github.com/rust-lang/stdarch.git
file:.git/config submodule.src/doc/book.active=true
---
---- [run-make] tests\run-make\pdb-buildinfo-cl-cmd stdout ----

error: rmake recipe failed to complete
status: exit code: 101
command: "C:\\a\\rust\\rust\\build\\x86_64-pc-windows-msvc\\test\\run-make\\pdb-buildinfo-cl-cmd\\rmake.exe"
--- stderr -------------------------------
thread 'main' panicked at C:\a\rust\rust\tests\run-make\pdb-buildinfo-cl-cmd\rmake.rs:28:5:
thread 'main' panicked at C:\a\rust\rust\tests\run-make\pdb-buildinfo-cl-cmd\rmake.rs:28:5:
assertion failed: &rfs::read("my_crate_name.pdb").find(br#""-C""#).is_some()
------------------------------------------




failures:
    [run-make] tests\run-make\pdb-buildinfo-cl-cmd

test result: FAILED. 261 passed; 1 failed; 90 ignored; 0 measured; 0 filtered out; finished in 83.40s

Some tests failed in compiletest suite=run-make mode=run-make host=x86_64-pc-windows-msvc target=x86_64-pc-windows-msvc
Build completed unsuccessfully in 1:30:46
make: *** [Makefile:102: ci-msvc-ps1] Error 1
  network time: Sat, 03 Aug 2024 04:32:34 GMT
##[error]Process completed with exit code 2.
Post job cleanup.
[command]"C:\Program Files\Git\bin\git.exe" version
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10224700120/job/28294215122)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/10224700120/job/28294215122","builder_name":"checks-actions"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #127877) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Oneirical commented:
I fixed the problem that caused the MSVC failure ("-C metadata" instead of "-Cmetatadata") and I have added the local helper function. Please run try again.

@rustbot review

jieyouxu commented:
@bors try

bors commented:
:hourglass: Trying commit 840a6612203c1972183a87f9886f7522eee89bc3 with merge 5a54c1e4665f3ed188d250ab1f953da5ec4172fd...
<!-- homu: {"type":"TryBuildStarted","head_sha":"840a6612203c1972183a87f9886f7522eee89bc3","merge_sha":"5a54c1e4665f3ed188d250ab1f953da5ec4172fd"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/10263906045/job/28402860129)
Build commit: 5a54c1e4665f3ed188d250ab1f953da5ec4172fd (`5a54c1e4665f3ed188d250ab1f953da5ec4172fd`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/10263906045/job/28402860129"},"merge_sha":"5a54c1e4665f3ed188d250ab1f953da5ec4172fd"} -->

jieyouxu commented:
Thanks!
@bors r+ rollup=iffy (pdb, pgo)

bors commented:
:pushpin: Commit 840a6612203c1972183a87f9886f7522eee89bc3 has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 840a6612203c1972183a87f9886f7522eee89bc3 -->
<!-- homu: {"type":"Approved","sha":"840a6612203c1972183a87f9886f7522eee89bc3","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

tgross35 commented:
@bors r-

Merge conflict, I don't know why bors didn't r- yet (it has been a few hours)

Oneirical commented:
Rebase complete, please re-approve. @rustbot ready

tgross35 commented:
@Oneirical github still says there is a conflict in `allowed_run_make_makefiles` (maybe another one merged?)

Oneirical commented:
> @Oneirical github still says there is a conflict in `allowed_run_make_makefiles` (maybe another one merged?)

Yes, it was caused by #128757 which just got merged.

Oneirical commented:
This is very weird, my Github branch [here](https://github.com/Oneirical/rust/blob/not-to-be-undertestimated/src/tools/tidy/src/allowed_run_make_makefiles.txt) is different from the [files changed](https://github.com/rust-lang/rust/blob/bf89c8c14bc97a4ab8bb8c9a567c241675bada35/src/tools/tidy/src/allowed_run_make_makefiles.txt) tab in this PR.

Note the `run-make/pgo-gen-lto/Makefile`.

Oneirical commented:
???

Oneirical commented:
I'm not too sure what happened there, but it seems like everything is back the way it should, and that the ghost merge conflict is gone.

Please re-approve.

tgross35 commented:
Maybe you rebased onto the local `master` without pulling first? In any case, looks like it worked now.

@bors r=jieyouxu

bors commented:
:pushpin: Commit 7d1a97fae9fc6d9f1790d054a40529fda4aacf3c has been approved by `jieyouxu`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=jieyouxu 7d1a97fae9fc6d9f1790d054a40529fda4aacf3c -->
<!-- homu: {"type":"Approved","sha":"7d1a97fae9fc6d9f1790d054a40529fda4aacf3c","approver":"jieyouxu","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 7d1a97fae9fc6d9f1790d054a40529fda4aacf3c with merge 86e7875c13f048ed03d413aa83215df94f0467e8...
<!-- homu: {"type":"BuildStarted","head_sha":"7d1a97fae9fc6d9f1790d054a40529fda4aacf3c","merge_sha":"86e7875c13f048ed03d413aa83215df94f0467e8"} -->

tgross35 commented:
This merged already, why did it get picked up?

@bors retry

compiler-errors commented:
@bors r-

tgross35 commented:
Bors didn't correctly mark a couple PRs as merged, isn't `r-`ing merge conflicts, and doesn't seem to be picking anything new up. I'm going to do a resync to see if it sorts things.

@bors treeclosed=100
